base/boot.jl

```diff
@@ -471,34 +488,116 @@ const NTuple{N,T} = Tuple{Vararg{T,N}}
+    if isdefined(self, :instance) && m === 0
+        self.instance
+    else
+        ccall(:jl_alloc_genericmemory, Ref{GenericMemory{isatomic,T}}, (Any, Int), self, m)
+    end
+(self::Type{GenericMemory{isatomic,T}})(::UndefInitializer, d::NTuple{1,Int}) where {T,isatomic} = self(undef, getfield(d,1))
+# empty vector constructor
+(self::Type{GenericMemory{isatomic,T}})() where {T,isatomic} = self(undef, 0)
+# copy constructors
+
+MemoryRef(mem::GenericMemory) = memoryref(mem)
+MemoryRef(ref::MemoryRef, i::Int) = memoryref(ref, i, true)
+MemoryRef(ref::MemoryRef, i::Int, check) = memoryref(ref, i, check)
+MemoryRef(mem::GenericMemory, i::Int) = memoryref(memoryref(mem), i, true)
+MemoryRef(mem::GenericMemory, i::Int, check) = memoryref(memoryref(mem), i, check)
+const Memory{T} = GenericMemory{:not_atomic, T}
+
+# construction helpers for Array
+new_as_memoryref(self::Type{MemoryRef{isatomic,T}}, m::Int) where {T,isatomic} = memoryref(fieldtype(self, :mem)(undef, m))
+
+# checked-multiply intrinsic function for dimensions
+_checked_mul_dims() = 1, false
+_checked_mul_dims(m::Int) = m, Intrinsics.ule_int(typemax_Int, m) # equivalently: (m + 1) < 1
+function _checked_mul_dims(m::Int, n::Int)
+    b = Intrinsics.checked_smul_int(m, n)
+    a = getfield(b, 1)
+    ovflw = getfield(b, 2)
+    ovflw = Intrinsics.or_int(ovflw, Intrinsics.ule_int(typemax_Int, m))
+    ovflw = Intrinsics.or_int(ovflw, Intrinsics.ule_int(typemax_Int, n))
+    return a, ovflw
+end
+function _checked_mul_dims(m::Int, d::Int...)
+   @_foldable_meta # the compiler needs to know this loop terminates
+   a = m
+   i = 1
+   ovflw = false
+   while Intrinsics.sle_int(i, nfields(d))
+     di = getfield(d, i)
+     b = Intrinsics.checked_smul_int(a, di)
+     ovflw = Intrinsics.or_int(ovflw, getfield(b, 2))
+     ovflw = Intrinsics.or_int(ovflw, Intrinsics.ule_int(typemax_Int, di))
+     a = getfield(b, 1)
+     i = Intrinsics.add_int(i, 1)
+   end
+   return a, ovflw
+end
+
+# convert a set of dims to a length, with overflow checking
+checked_dims() = 1
+checked_dims(m::Int) = m # defer this check to Memory constructor instead
+function checked_dims(d::Int...)
+    b = _checked_mul_dims(d...)
+    getfield(b, 2) && throw(ArgumentError("invalid Array dimensions"))
+    return getfield(b, 1)
+end
+
+# bare array constructor from MemoryRef and size
+#=eval(Core, :(function (self::Type{Array{T,N}})(ref::MemoryRef, dims::NTuple{N,Int}) where {N, T}
+    length = checked_dims(dims...)
+    length === 0 || memoryref(self, length, true) # make sure last element is inbounds
+    if !(memoffset(ref) === 0 && length === ref.mem.length
+        ref = memoryref(ccall(:jl_genericmemory_slice, Memory{T}, (Memory{T}, Int), ref.mem, memoffset(ref) length)
+    end
+    return $(Expr(:new, :self, :(ref), :(dims)))
+end))=#
+eval(Core, :(function (self::Type{Array{T,1}})(ref::MemoryRef, dims::Tuple{Int}) where {T}
```

Comment:
These evals are because.....we're using `Expr(:new, ...)`? And because we don't have normal constructors available for `Array` since it's /src-defined?

Reply:
yes